### PR TITLE
FormTemplate Versions Option 1 - Hash

### PR DESCRIPTION
**Pull request by ailZhou**
### Description

Using hash to determine which version of the formTemplate to return to the user.

When a user creates a form, the system will get the base WP/SAR json, make a copy of it and then run the scan and transformation rules on the form copy. Once it finishes modifying that form, it will generate a hash out of that data and then check the database to see if a formTemplate with the same hash exist. If that hash exist, we retrieve that saved formTemplate and return it to the user, if not, we save the form copy we just created to the database. 

This is a temporary fix for the bug in prod until we can fully implement @benmartin-coforma FormTemplate version.

Screenshot 2024-01-09 at 4 44 58 PM

### Related ticket(s)

CMDCT-3161

---
### How to test

1) Run local dynamodb to be able to see if the versions are updating correctly

1) Sign into MFP as a state user.
2) Create, fill out and submit a WP.
3) Sign out and sign in as an Admin. Approve the submitted WP
4) Sign out and sign back in as the previous state user.
5) Create a WP copy.
6) Do the previous steps as many times as deserved. Each WP copied should be its own version in the database.
7) Sign in as a different state user and go through the same process
8) This state user should be using the previous created forms instead of creating a new form.

### Important updates

N/A

---
### Author checklist

- [x] I have performed a self-review of my code
- [x] I have added [thorough](https://shorturl.at/aejkF) tests, if necessary
- [ ] I have updated relevant documentation, if necessary
---

_convert to a different template: [test → val](?expand=1&template=test-to-val-deployment.md)_ | _[val → prod](?expand=1&template=val-to-prod-deployment.md)_

### CODE REVIEW

**Review by gmrabian — APPROVED** [review]
looks good! works as described. Nice job!

File: services/app-api/handlers/reports/create.ts [hunk 1]
```diff
@@ -143,8 +143,13 @@ export const createReport = handler(
...
 
     if (overrideCopyOver) {
+      if (unvalidatedMetadata?.copyReport?.reportPeriod)
+        reportPeriod = unvalidatedMetadata?.copyReport?.reportPeriod;
+      if (unvalidatedMetadata?.copyReport?.reportYear)
+        reportYear = unvalidatedMetadata?.copyReport?.reportYear;
+
       reportYear = reportPeriod == 2 ? reportYear + 1 : reportYear;
-      reportPeriod = reportPeriod == 1 ? 2 : 1;
+      reportPeriod = (reportPeriod % 2) + 1;
```
[the remark this answers is not in the document]
Comment (karla-vm) [reply services/app-api/handlers/reports/create.ts:152]: seconded!

**Review by karla-vm — APPROVED** [review]

File: services/app-api/utils/formTemplates/formTemplates.test.ts [hunk 2]
```diff
@@ -17,13 +19,44 @@ import {
 } from "../testing/setupJest";
 import s3Lib from "../s3/s3-lib";
 import dynamodbLib from "../dynamo/dynamodb-lib";
+import { AnyObject } from "yup/lib/types";
```
[the remark this answers is not in the document]
Comment (ailZhou) [reply services/app-api/utils/formTemplates/formTemplates.test.ts]: fixed, thanks for catching this!

**Review by britt-mo — APPROVED** [review]
Looks good, used your branch with Nick on our current in progress work, everything id functioning correctly

**Review by cassandradanger — APPROVED** [review]
tested in the browser and I think everything looks good!!

**Pull request merged**